core/embed/trezorhal/stm32u5/secret.c

```diff
@@ -199,14 +226,53 @@ secbool secret_optiga_get(uint8_t dest[SECRET_OPTIGA_KEY_LEN]) {
   return res;
 }
 
-void secret_optiga_hide(void) {
+// Deletes the optiga pairing secret from the register
+static void secret_optiga_uncache(void) {
   volatile uint32_t *reg1 = &TAMP->BKP8R;
   for (int i = 0; i < 8; i++) {
-    *reg1 = 0;
-    reg1++;
+    reg1[i] = 0;
   }
 }
+#endif
 
 void secret_erase(void) {
   ensure(flash_area_erase(&SECRET_AREA, NULL), "secret erase");
 }
+
+void secret_show_install_restricted_screen(void) {
+  // this should never happen on U5
+  __fatal_error("INTERNAL ERROR", "Install restricted", __FILE__, __LINE__,
+                __func__);
+}
+
+void secret_prepare_fw(secbool allow_run_with_secret, secbool trust_all) {
+  /**
+   * The BHK is copied to the backup registers, which are accessible by the SAES
+   * peripheral. The BHK register is locked, so the BHK can't be accessed by the
+   * software.
+   *
+   * When optiga is paired, pairing secret is copied to the backup registers
+   * and access to the secret storage is disabled. Otherwise, access to the
+   * secret storage kept to allow optiga pairing in prodtest.
+   *
+   * Access to the secret storage is disabled for non-official firmware in
+   * all-cases.
+   */
+
+  secret_bhk_load();
+  secret_bhk_lock();
+#ifdef USE_OPTIGA
+  if (sectrue == allow_run_with_secret && sectrue == secret_optiga_present()) {
+    secret_optiga_cache();
+    secret_disable_access();
+  } else {
+    secret_optiga_uncache();
+  }
+#else
+  secret_disable_access();
+#endif
+
+  if (sectrue != trust_all) {
+    secret_disable_access();
+  }
```

Comment:
This means that if `allow_run_with_secret` is false and `trust_all` is true, then the firmware will have access to the secret area? I think we don't have any header with this kind of configuration, but the logic seems quite clearly wrong. Also `trust_all` seems quite irrelevant here. I am thinking it should be something like this:

```suggestion
#ifdef USE_OPTIGA
  secret_optiga_uncache();
  if (sectrue == allow_run_with_secret) {
    if (secfalse != secret_optiga_present()) {
      secret_optiga_cache();
      secret_disable_access();
    }
  } else {
    secret_disable_access();
  }
#else
  secret_disable_access();
#endif
```

Reply:
Hm, i am a bit reluctant about the removal of `trust_all`. The `VTRUST_SECRET` was more meant for the optiga pairing secret, and here on U5 the secret sector contains also bootloader downgrade protection, which i think we never want available for firmwares without `trust_all`. Given the existing combinations of flags in vendor header this is theoretical, but i think stricter checks are better. 

The difference would be that if we ever have vendor header with `allow_run_with_secret=true` and no `trust-all`, then before the optiga secret if written to the flash, firmware with this vendor header could downgrade the bootloader.

Reply:
The rest of the logic is changed here https://github.com/trezor/trezor-firmware/commit/1792b4f1d928a46e731505df40f5b4ab9ede5bfb

Reply:
I dont think it would be great if prodtest had access to the secret sector always, regardles of pairing secret being present or not - the binary can leak any time thus its no better than official firmware having access.

The second argument is more interesting, not entirely sure its worth it